packages/request-handler/src/index.js

```diff
@@ -0,0 +1,225 @@
+      log.debug('Handing queue records', { count: records.length });
+
+      const results = await Promise.allSettled(
+        records.map((qr) => this.handleQueueRecord(context, qr, queueClient, log)),
+      );
+
+      const batchItemFailures = results
+        .filter((r) => r.status === 'rejected')
+        .map((fail) => fail.reason)
+        .map((itemIdentifier) => ({ itemIdentifier }));
+      if (batchItemFailures.length === 0) {
+        log.debug('All batch items processed successfully');
+        return new Response();
+      } else {
+        log.info('Failed to process batch items', {
+          items: batchItemFailures,
+          count: batchItemFailures.length,
+        });
+        return new Response(
+          JSON.stringify({
+            batchItemFailures,
+          }),
+          { status: 206 },
+        );
+      }
+    } else {
+      log.debug('Handing POST payload');
+      const event = helper.extractOriginalEvent();
+      return this.handleEvent(event, context);
+    }
+  }
+
+  /**
+   * Handles a single event
+   * @param {Record<string,any>} event the event to handle
+   * @param {import('@adobe/content-lake-commons').contextHelper.UniversalishContext} context
+   *  the current context
+   * @returns {Promise<Response>} the response from handling the event
+   */
+  async handleEvent(event, context) {
+    const { action } = event || {};
+    if (!action) {
+      throw new RestError(400, 'Missing parameter [action]');
+    }
+    if (!this.#handlers[action]) {
+      throw new RestError(400, `Invalid action [${action}]`);
+    }
+    return this.#handlers[action](event, context);
+  }
+
+  /**
+   * Handles a queue event record
+   * @param {import('@adobe/content-lake-commons').contextHelper.UniversalishContext} context
+   * @param {import('@adobe/content-lake-commons').QueueRecord} record
+   * @param {QueueClient} queueClient
+   * @param {import('@adobe/content-lake-commons').contextHelper.Logger} log
+   * @returns {Promise<void>}
+   */
+  async handleQueueRecord(context, record, queueClient, log) {
+    try {
+      log.debug('Handling queue record', {
+        record,
+      });
+      const event = await queueClient.readMessageBody(record.body);
+      const res = await this.handleEvent(event, context);
+      if (res.ok) {
+        log.debug('Record handled successfully, removing from queue', {
+          record,
+        });
+        await queueClient.removeMessage(record.receiptHandle);
+      } else {
+        log.warn('Record handling unsuccessful', {
+          record,
+          res,
+        });
+        throw record.messageId;
+      }
+    } catch (err) {
+      log.warn('Failed to handle queue record', { record, err });
+      throw record.messageId;
+    }
+  }
+
+  /**
+   * Adds an additional wrap function to the invocation
+   * @param {import("@adobe/helix-shared-wrap").WrapFunction} fn
+   */
+  withWrapFunction(fn) {
+    this.#wrapFn = fn;
```

Comment:
nitpick: do we want any security check to make sure it's a function?

Reply:
I mean, it's an internal feature, you should find out pretty quickly that it doesn't work if you pass in something that's not a function.